File: codeforces/2093/E_Min_Max_MEX.cpp
#include <bits/stdc++.h>
using namespace std;

using i64 = long long;

int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);

  int t;
  cin >> t;

  while (t--) {
    i64 n, k;
    cin >> n >> k;

    vector<i64> a(n), vals;
    for (i64 &x : a) {
      cin >> x;
      vals.push_back(x);
    }

    sort(vals.begin(), vals.end());
    vals.erase(unique(vals.begin(), vals.end()), vals.end());

    i64 mex = 0;
    for (i64 v : vals) {
      if (v == mex)
        ++mex;
      else if (v > mex)
        break;
    }

    map<i64, int> cnt;
    for (const i64 &x : a) {
      cnt[x]++;
    }

    i64 left = 0, right = mex, ans = 0;

    while (left <= right) {
      i64 mid = (left + right) / 2;
      bool valid = true;

      if (mid > 0) {
        for (i64 i = 0; i < mid; ++i) {
          auto it = cnt.find(i);
          if (it == cnt.end() || it->second < k) {
            valid = false;
            break;
          }
        }
        if (!valid) {
          right = mid - 1;
          continue;
        }
      }

      if (mid == 0) {
        if (k <= n) {
          ans = max(ans, 0LL);
          left = mid + 1;
        } else {
          right = mid - 1;
        }
      } else {
        i64 req = mid;
        vector<int> freq(req, 0);
        i64 c_seg = 0, col = 0;

        for (const i64 &x : a) {
          if (x >= 0 && x < req) {
            if (++freq[x] == 1) ++col;
            if (col == req) {
              ++c_seg;
              fill(freq.begin(), freq.end(), 0);
              col = 0;
            }
          }
        }

        if (c_seg >= k) {
          ans = max(ans, mid);
          left = mid + 1;
        } else {
          right = mid - 1;
        }
      }
    }

    cout << ans << '\n';
  }

  return 0;
}

File: codeforces/2093/E_Min_Max_MEX.py
import sys
from collections import defaultdict, Counter
from random import shuffle

input = sys.stdin.readline

t = int(input())

for _ in range(t):
    n, k = map(int, input().split())
    a = list(map(int, input().split()))
    mex = 0
    a_cp=a.copy()
    shuffle(a_cp)
    s = set(a_cp)
    while mex in s:
        mex += 1

    cnt = Counter(a)

    left = 0
    right = mex
    ans = 0

    while left <= right:
        mid = (left + right) // 2

        valid = True
        if mid > 0:
            for i in range(mid):
                if i not in s or cnt[i] < k:
                    valid = False
                    break
            if not valid:
                right = mid - 1
                continue

        if mid == 0:
            if k <= n:
                ans = 0
                if ans > ans:
                    ans = ans
                left = mid + 1
            else:
                right = mid - 1
        else:
            req = mid
            c_cnt = defaultdict(int)
            c_seg = 0
            col = 0
            for num in a:
                if 0 <= num < req:
                    if c_cnt[num] == 0:
                        col += 1
                    c_cnt[num] += 1
                    if col == req:
                        c_seg += 1
                        c_cnt.clear()
                        col = 0
            if c_seg >= k:
                if mid > ans:
                    ans = mid
                left = mid + 1
            else:
                right = mid - 1
    print(ans)
